### PR TITLE
Implement CO to set all QuickEffectRack Effects to the same Chain.

### LINKED ISSUE #11371: Common Mixer `loaded_chain_preset` CO
### Feature Description

See https://github.com/mixxxdj/mixxx/pull/4027#discussion_r1136877275

It would make sense for to have a single CO that changes the `loaded_chain_preset` for all Channels of the mixer since that workflow is common on most controllers.

### PR DESCRIPTION
**Pull request by Swiftb0y**
Fixes #11371 

Current Issues to discuss:
- [ ] `[QuickEffectRack1]` is a new CO group, not sure if thats desired or where else to put the CO.
- [x] `[QuickEffectRack1]`,`loaded_chain_preset` is zero-based while their `[QuickEffectRack1_[ChannelN]]` equivalent is 1-based. Considering the first real preset starts at 2 now (since 1 is the empty chain introduced in #10859). 
  - [x] When making the CO 0-based again ,`[EqualizerRack1]` and `[EffectRack1]` `loaded_chain_preset` now get initialized to weird, very large values for some reason.

### CODE REVIEW

**Comment by Swiftb0y** [comment]
Quick note: #10859 introduced the inconsistency that `QuickEffectRack1_[ChannelN]` `loaded_chain_preset` is now offset by one compared to `[EffectRack1_EffectUnitN]` because of the prepended empty preset.

File: src/effects/effectchain.cpp [hunk 1]
```diff
@@ -340,8 +340,7 @@ void EffectChain::slotControlChainPresetSelector(double value) {
 }
 
 void EffectChain::slotControlLoadedChainPresetRequest(double value) {
-    // subtract 1 to make the ControlObject 1-indexed like other ControlObjects
-    int index = static_cast<int>(value) - 1;
+    int index = static_cast<int>(value);
```
Comment (ronso0) [inline src/effects/effectchain.cpp:343]: Why not add `QuickEffectChain::slotControlLoadedChainPresetRequest(double value)` so the index is stll 1-based for standard chains and 0-based for QuickEffects?

We just need to add a note to the documentation https://github.com/mixxxdj/manual/issues/501

**Comment by daschuer** [comment]
I will move this PR to 2.5, because it is a feature.

**Comment by Swiftb0y** [comment]
I'm still unhappy about https://github.com/mixxxdj/mixxx/pull/11378#issuecomment-1494798902

**Comment by ronso0** [comment]
Yeah, that's a pity, sorry for that!

I'm checking whether it's feasible to fix the index issue while avoiding the issue with standard effect chains I described in https://github.com/mixxxdj/mixxx/pull/10859#discussion_r962314889

**TLDR - the dilemma:**
You can select `---` in a any chain to clear all slots.
Since all effect chains provide controls to load effects, the actually write-protected chain preset can be modified (saving won't work though).
Hence you can have the '---' preset loaded and effects loaded, which renders the '---' item useless.

Actually we can run into this situation with any chain, but it's problematic only for the standard chains due to how their states are saved to effects.xml. We save
* preset name
* states of all parameters of all slots

So when restoring standard chains on startup, we restore the previous state, but also select the '---' item, hence it's useless again.

(for QuickEffect and EQ chains it's all good: only the preset name is stored, so if the '---' preset has been modified in one of these chains we'd now load the desired empty preset)

**Workaround (dirty)**
* prepend '---' to all preset lists
* catch EffectSlot::effectChanged
* if '---' is loaded and the slot is not empty:
   * load (nameless) empty preset (initial state of standard chains)
   * reload the selected effect
   * all good again, '---' is unloaded and can be used to clear the chain

**Workaround (somewhat clean)**
* migrate the chain state to a new, nameless preset, preferrably with a new `EffectChainPresetManager` method
* = do all of the above, just without the effect unload/reload hack

I'm working on it, will try to revert existing / not introduce more hacks/inconsistencies.
Maybe fast and clean enough to integrate it into 2.4.0

**Comment by ronso0** [comment]
> When making the CO 0-based again ,[EqualizerRack1] and [EffectRack1] loaded_chain_preset now get initialized to weird, very large values for some reason.

4.29497e+09 to be precise ; )

int -> uint wrap-around?
`EffectChain::presetIndex()` gets `int` from
`EffectChainPresetManager::quickEffectPresetIndex(presetName)`
but `EffectChain::setControlLoadedPresetIndex` expects `uint`.

I have WIP where I add the `---` preset to the common chain preset list and make `loaded_chain_preset` 0-indexed, so we get
**QuickEffect & standard effect chains:**
* unknown preset: -1 (default with clean profile)
* know presets incl. `---: 0 - end

**EQ chains + main output chain:**
* always -1 because we don't load presets, only an effect into slot 1

How to proceed?
My WIP at least that restores consistency, while avoiding the `---` save/restore issue I described above.
I'll share it soonish.

**Comment by ronso0** [comment]
fix is in #12561

**Comment by ronso0** [comment]
@Swiftb0y Want rebase onto main and adjust the description?

File: src/effects/effectsmanager.cpp [hunk 2]
```diff
@@ -48,6 +55,20 @@ EffectsManager::EffectsManager(
...
+            [this](double index) {
+                if (index < 0) {
+                    return;
+                }
+                for (QuickEffectChainPointer quickEffect :
+                        std::as_const(m_quickEffectChains)) {
+                    quickEffect->loadChainPreset(quickEffect->presetAtIndex(
+                            static_cast<int>(index)));
+                }
+            });
```
Comment (Swiftb0y) [inline src/effects/effectsmanager.cpp:71]: note to self, extract into member func for better formatting.

File: src/effects/effectsmanager.cpp [hunk 3]
```diff
@@ -170,6 +191,24 @@ void EffectsManager::addQuickEffectChain(const ChannelHandleAndGroup& deckHandle
...
+                                [presetIndex](const QuickEffectChainPointer
+                                                quickEffectChain) {
+                                    return presetIndex ==
+                                            quickEffectChain->presetIndex();
+                                });
+                m_unifiedQuickEffectPresetIndex.set(
+                        allQuickEffectChainsHaveSameIndex
+                                ? presetIndex
+                                : kMixedQuickEffectsSelected);
+            });
```
Comment (Swiftb0y) [inline src/effects/effectsmanager.cpp:211]: same here.

**Comment by Swiftb0y** [comment]
yikes. tags galore. I don't know when I'll be able to properly pick this up. Feel free to adopt.
